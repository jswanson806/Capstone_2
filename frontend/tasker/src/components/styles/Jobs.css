.jobs-header-container {
    text-align: center;
    margin: 2% 2%;
}

.jobs-buttons-container {
    padding-bottom: 2rem;
    text-align: center;
}

.jobs-button {
    margin: 2px 5px;
    padding: 5px;
}

.jobs-jobCard-container {
    justify-content: center;
    padding: 20px;
}

.jobs-notFound-message {
    display: flex;
    justify-content: center;
    align-items: center;
    height: 10vh;
}

.jobs-dynamic-page-header {
    font-weight: bold;
    
}

.jobCard-card {
    height: 15em;
}

.jobCard-badge {
    position: static !important;
    padding: 5px !important;
}

.jobCard-badge-container {
  position: relative;
  top: 34%;
}

p {
    position: relative;
    top: 25%;
}

h5 {
    position: relative;
    top: 25%;
}

.item {
    flex: 1 0 20%;
    max-width: 20%;
}