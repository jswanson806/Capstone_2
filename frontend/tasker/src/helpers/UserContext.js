import React, { createContext, useState } from 'react';

// Create the context
const UserContext = createContext();

// Create a provider for the context
const UserProvider = ({ children }) => {
  const [user, setUser] = useState(localStorage.getItem('user') || '');

  // Function to update the user in the state and localStorage
  const updateUser = (newUser) => {
    setUser(newUser);
    localStorage.setItem('user', newUser);
  };

  return (
    <UserContext.Provider value={{ user, updateUser }}>
      {children}
    </UserContext.Provider>
  );
};

export { UserContext, UserProvider };
