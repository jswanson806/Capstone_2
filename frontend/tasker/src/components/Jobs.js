import React, { useContext, useEffect, useState } from 'react';
import TaskerApi from '../api';
import JobCard from './JobCard.js';
import { UserContext } from '../helpers/UserContext';
import CreateJob from './CreateJob.js';
import {
    Spinner, 
    Button, 
    Modal,
    ModalHeader,
    ModalBody,
    Row,
    Container
} from "reactstrap";
import "./styles/Jobs.css";

const Jobs = () => {

    const jobCardsInitialState = [];
    const jobsInitialState = '';

    const {user} = useContext(UserContext)

    const [jobs, setJobs] = useState(jobsInitialState);
    const [jobCards, setJobCards] = useState(jobCardsInitialState);
    const [currUser, setCurrUser] = useState(null);
    const [header, setHeader] = useState('');
    const [buttons, setButtons] = useState('');
    const [isLoading, setIsLoading] = useState(true);
    const [triggerEffect, setTriggerEffect] = useState(false);
    const [notFound, setNotFound] = useState(null);

    // view toggles
    const [isWorker, setIsWorker] = useState(false);
    const [showCreateJob, setShowCreateJob] = useState(false);
    
    /** On initial render
     * 
     *  Calls functions to set state of currUser
     *  Gets the current user from the database
     */
    useEffect(() => {
        // parsed user string received from user context
        const parsedUser = JSON.parse(user);
        // call api to get user from db and set state of currUser
        getCurrUser(parsedUser.id);
        
    }, []);

    
    /** Sets the default content for the page depending on isWorker boolean from the currUser state
     * 
     * Calls API for initial list of jobs
     */
    useEffect(() => {
        
        // if currUser is populated & create job form is not shown
        if(currUser && !showCreateJob){

            if(!isWorker){ // if user is not worker, get the jobs posted by this user
                getAndSetActiveUserJobs(currUser.id);
                // set buttons for users
                setButtons(renderUserButtons());
                // set header for page
                renderHeader('My Active Jobs');
            } else { // else the user is a worker, get all available jobs
                getAndSetAllAvailableJobs();
                // set buttons for workers
                setButtons(renderWorkerButtons());
                // set header for page
                renderHeader('Available Jobs');
            }
        }

    }, [
        currUser,
        isWorker,
        triggerEffect
    ]);


    /** Calls functions to create and set job cards, state of notFound, and set state of isLoading
     *  
     * Waits for jobs and currUser to be populated
     * 
    */
    useEffect(() => {

        // only call once jobs and currUser have been populated
        if(jobs !== jobsInitialState && currUser) {
            // create the job cards
            createJobCards();
            // update state of notFound
            setNotFound(false);
        } else { // jobCards state is default
            // update state of notFound
            setNotFound(true);
        }
        
    }, [ jobs, currUser ]);

    /** Updates state of isLoading once state of notFound is updated */
    useEffect(() => {
        // notFound can either be truthy or falsy to trigger isLoading state update
        if(notFound || notFound === false){
            // set isLoading false
            setIsLoading(false);
        }
    }, [notFound])


    /**
    * Fetches data for a single user from the Tasker API based on the given ID.
    *
    * Updates state of currUser with the response object
    * Updates isWorker with the isWorker property of the response object
    */
    async function getCurrUser(id){
        try {
            const res = await TaskerApi.getSingleUser(id);
            const { user } = res;
            setCurrUser(user);
            setIsWorker(user.isWorker);
        
        } catch(err) {
            console.error(err);
        }
    }

    /**
    * Sets state of JobCards based on current view toggles.
    * 
    * Uses filtering based on the current user, view toggle states, and existing filtered jobs to map JobCard components to jobCards state
    * 
    */
    async function createJobCards(){
        // convert user applications to Set
        const applications = new Set(currUser.applications);
        // map each job to a JobCard componenet
        const jobCards = jobs.map((job) => {
            return (
                
                    <JobCard
                        user={currUser}
                        applications={applications}
                        job={job}
                        getCurrUser={getCurrUser}
                        setJobs={setJobs}
                        key={job.id}
                        triggerEffect={() => setTriggerEffect(!triggerEffect)}
                        data-testid="jobCard-component"
                    />
                
            );
        });

        setJobCards(jobCards);
    };

    /** 
     * Fetches data for jobs posted by the user where the job status is 'active'
     * 
     * Calls api to findInProgressUserJobs
     * 
     * Updates state of jobs with the response array of jobs
     */
    async function getAndSetActiveUserJobs(userId){
        setIsLoading(true);
        try {
            const inProgressUserJobs = await TaskerApi.getActiveUserJobs(userId);
            if(!inProgressUserJobs.jobs) {
                setJobs(jobsInitialState);
                setJobCards(jobCardsInitialState);
            } else {
                const { jobs } = inProgressUserJobs;
                setJobs(jobs);
            };
        } catch(err) {
            console.error(err);
        }
    }

    /**
     * Fetches data for jobs posted by the user where the job status is 'pending review'
     * 
     * Calls api to findPendingReviewUserJobs
     * 
     * Updates state of jobs with the response array of jobs
    */
    async function getAndSetPendingReviewUserJobs(userId){
        setIsLoading(true);
        try {
            const allPendingReviewUserJobs = await TaskerApi.getPendingReviewUserJobs(userId);
            if(!allPendingReviewUserJobs.pendingReviewUserJobs) {
                setJobs(jobsInitialState);
                setJobCards(jobCardsInitialState);
            } else {   
                const { pendingReviewUserJobs } = allPendingReviewUserJobs;
                setJobs(pendingReviewUserJobs);
            }
        } catch(err) {
            console.error(err);
        }
    }

    /**
     * Fetches data for jobs available for applications by workers
     * 
     * Calls api to findAllAvailableJobs
     * 
     * Updates state of jobs with the response array of jobs
    */
    async function getAndSetAllAvailableJobs(){
        setIsLoading(true);
        try {
            const allAvailableJobs = await TaskerApi.getAllAvailableJobs();
            const { allJobs } = allAvailableJobs;
            setJobs(allJobs);
        } catch(err) {
            console.error(err);
        }
    }

    /**
     * Fetches data for jobs to which worker has already applied
     * 
     * Calls api to findAllAppliedWorkerJobs
     * 
     * Updates state of jobs with the response array of jobs
    */
    async function getAndSetAllAppliedWorkerJobs(workerId){
        setIsLoading(true);
        try {
            const allAppliedJobs = await TaskerApi.getAppliedWorkerJobs(workerId);
            if(!allAppliedJobs.appliedJobs){
                setJobs(jobsInitialState);
                setJobCards(jobCardsInitialState);
            } else {
                const { appliedJobs } = allAppliedJobs;
                setJobs(appliedJobs);
            }
        } catch(err) {
            console.error(err);
        }
    }

    /**
     * Fetches data for jobs to which worker has already been assigned
     * 
     * Calls api to findAllInProgressWorkerJobs
     * 
     * Updates state of jobs with the response array of jobs
    */
    async function getAndSetAllAssignedWorkerJobs(workerId){
        setIsLoading(true);
        try {
            const allAssignedWorkerJobs = await TaskerApi.getAllAssignedWorkerJobs(workerId);
            if(!allAssignedWorkerJobs.assignedJobs){
                setJobs(jobsInitialState);
                setJobCards(jobCardsInitialState);
            } else {
                const { assignedJobs } = allAssignedWorkerJobs;
                setJobs(assignedJobs);
            }
        } catch(err) {
            console.error(err);
        }
    }
 
    const toggleCreateJob = () => {
        setShowCreateJob(!showCreateJob);
    }

    const renderHeader = (pageHeader) => {
        setHeader(<h2 className="jobs-dynamic-page-header">{pageHeader}</h2>)
    }


    const renderWorkerButtons = () => { // buttons for workers
        return (
            <div className="jobs-button-container">
                
                <Button onClick={() => {
                        getAndSetAllAvailableJobs();
                        renderHeader('Available Jobs');
                    }}
                    className="jobs-button"
                    color="info"
                    data-testid="available-jobs-menu-item">
                    Available Jobs
                </Button>
                <Button onClick={() => {
                        getAndSetAllAppliedWorkerJobs(currUser.id);
                        renderHeader('My Applications');
                    }}
                    className="jobs-button"
                    color="info" 
                    data-testid="applications-menu-item">
                    Applications
                </Button>
                <Button onClick={() => {
                        getAndSetAllAssignedWorkerJobs(currUser.id);
                        renderHeader('My Assigned Jobs');
                    }} 
                    className="jobs-button"
                    color="info"
                    data-testid="my-jobs-menu-item">
                    My Jobs
                </Button>

            </div>
        );
    };
        
    const renderUserButtons = () => { // buttons for users
        return (
            <div className="jobs-button-container">
               
                <Button onClick={() => {
                        getAndSetActiveUserJobs(currUser.id);
                        renderHeader('My Active Jobs');
                    }} 
                    className="jobs-button"
                    color="info"
                    data-testid="my-jobs-menu-item">
                    Active Jobs
                </Button>
                <Button onClick={() => {
                        getAndSetPendingReviewUserJobs(currUser.id);
                        renderHeader('Jobs Pending Review');
                    }} 
                    className="jobs-button"
                    color="info"
                    data-testid="pending-review-menu-item">
                    Pending Review
                </Button>
                <Button 
                    className="jobs-button" 
                    color="info" 
                    data-testid="jobs-create-job-button" 
                    onClick={toggleCreateJob}
                >
                    Create New Job
                </Button>
            </div>
        );
    };


    // default state of page shows spinner when isLoading is true
    if(isLoading){
        return (
        <div className="spinner">
            <Spinner>
                Loading...
            </Spinner>
        </div>
          )
    }

    return (
        <div className="jobs-container">

            <div className="jobs-header-container">
                {header}
            </div>

            
                <div className="jobs-buttons-container">
                    {buttons}
                </div>
            
                
            
            
                <CreateJob
                    isOpen={showCreateJob} 
                    toggle={toggleCreateJob}
                    onCreate={() => 
                    {
                        setTriggerEffect(!triggerEffect);
                        toggleCreateJob();
                    }}
                    onClose={() => 
                    {
                        toggleCreateJob();
                    }}
                /> 
            
            
            
                <Container>
                    <Row 
                        xs={1} 
                        sm={2} 
                        md={3} 
                        lg={4} 
                        xl={5} 
                        className="justify-content-center"
                    >
                        {jobCards}
                    </Row>
                </Container>
            

            {notFound && (
                <span className="jobs-notFound-message">It looks like no jobs were found.</span>
            )}
        </div>
    )
}

export default Jobs;