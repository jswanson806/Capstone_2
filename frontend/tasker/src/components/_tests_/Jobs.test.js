import React from 'react';
import { render, act, screen, waitFor, fireEvent } from '@testing-library/react';
import Jobs from '../Jobs.js';
import { UserContext } from '../../helpers/UserContext.js';
import TaskerApi from '../../api.js';
// import JobDetails from '../JobDetails.js';


const userValue = { user: '{"id": 1, "email": "test@email.com", "isWorker": true}' };
const userValue2 = { user: '{"id": 1, "email": "test@email.com", "isWorker": false}' };

const getSingleUser = jest.spyOn(TaskerApi, 'getSingleUser');
const getAllAvailableJobs = jest.spyOn(TaskerApi, 'getAllAvailableJobs');
const getAppliedWorkerJobs = jest.spyOn(TaskerApi, 'getAppliedWorkerJobs');
const getAllAssignedWorkerJobs = jest.spyOn(TaskerApi, 'getAllAssignedWorkerJobs');
const getActiveUserJobs = jest.spyOn(TaskerApi, 'getActiveUserJobs');
const getPendingReviewUserJobs = jest.spyOn(TaskerApi, 'getPendingReviewUserJobs');
const getBeforeImage = jest.spyOn(TaskerApi, 'getBeforeImage');

describe('smoke and snapshot tests', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: true, applications: [2]}});
        getAllAvailableJobs.mockResolvedValueOnce({allJobs: [
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'pending',
                applicants: [1]
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });
    
    test('Jobs component renders correctly', async () => {

        await act(async () => {render(
                    <UserContext.Provider value={userValue}>
                        <Jobs />
                    </UserContext.Provider>
        )
        });

    });

    it("should match snapshot", async() => {

        expect.assertions(1)

        let asFragment;

        await act(async () => {const {asFragment: fragment} = render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        asFragment = fragment;
        });

        expect(asFragment()).toMatchSnapshot();
    })

});



describe('tests the useEffect hooks upon render and data retrieval for workers', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: true, applications: [2]}});
        getAllAvailableJobs.mockResolvedValueOnce({allJobs: [
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'pending',
                applicants: [1]
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });
    
    test('Jobs hooks - first useEffect calls api to retrieve worker', async () => {

        expect.assertions(1);

        await act(async () => {render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        expect(getSingleUser).toHaveBeenCalledTimes(1);

    });

    test('Jobs hooks - second useEffect calls api to retrieve jobs', async () => {

        expect.assertions(2);

        await act(async () => {render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        expect(getSingleUser).toHaveBeenCalledTimes(1);
        expect(getAllAvailableJobs).toHaveBeenCalledTimes(1);

    });

    test('Jobs hooks - third useEffect creates jobCard(s) and sets loading to false', async () => {

        expect.assertions(2);

        await act(async () => {render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        await waitFor(() => {
            const loadingSpinner = document.querySelector('.spinner');
            const jobCard = document.querySelector('.jobCard-card');
            expect(loadingSpinner).not.toBeInTheDocument();
            expect(jobCard).toBeInTheDocument();
        })
    });

});

describe('tests the useEffect hooks upon render and data retrieval for users', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: false}});
        getActiveUserJobs.mockResolvedValueOnce({jobs: [
            {
                id: 1, 
                title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                postedBy: 1, 
                status: 'active',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'pending',
                applicants: [1]
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });
    
    test('Jobs hooks - first useEffect calls API to retrieve user', async () => {

        expect.assertions(1);

        await act(async () => {render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        expect(getSingleUser).toHaveBeenCalledTimes(1);

    });

    test('Jobs hooks - second useEffect calls API to retrieve jobs', async () => {

        expect.assertions(1);

        await act(async () => {render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        expect(getActiveUserJobs).toHaveBeenCalledTimes(1);

    });

    test('Jobs hooks - third useEffect creates jobCard(s)', async () => {

        expect.assertions(1);

        await act(async () => {render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        await waitFor(() => {
            const jobCard = document.querySelector('.jobCard-card');
            expect(jobCard).toBeInTheDocument();
        })

    });

});


describe('tests the createJobCards function following api calls to retrieve job data for workers', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: true, applications: [2]}});
        getAllAvailableJobs.mockResolvedValueOnce({allJobs: [
            {
                id: 1, 
                title: 'testing jobs 1',
                body: 'job 1 body for testing the job card',
                posted_by: 2,
                status: 'active',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2',
                body: 'job 2 body',
                posted_by: 1,
                status: 'pending',
                applicants: [1]
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });

    test('applications button calls api to get all jobs to which worker has applied and creates job cards from returned data', 
        async () => {

            expect.assertions(5);

            let renderResult;

            await act(async () => {renderResult = render(
                <UserContext.Provider value={userValue}>
                    <Jobs />
                </UserContext.Provider>
            )
            });
            const { container } = renderResult;

            expect(getAllAvailableJobs).toHaveBeenCalledTimes(1);

            const btn = screen.queryByTestId('applications-menu-item')
            expect(btn).toBeInTheDocument();


            await act(async () => {
                getAppliedWorkerJobs.mockResolvedValueOnce({appliedJobs: [
                    {
                        id: 2,
                        title: 'testing jobs 2',
                        body: 'job 2 body',
                        postedBy: 1,
                        status: 'pending',
                        applicants: [1]
                    }
                ]});

                fireEvent.click(btn)
            })


            expect(getAppliedWorkerJobs).toHaveBeenCalledTimes(1);


            expect(container).not.toHaveTextContent('testing jobs 1');
            expect(container).toHaveTextContent('testing jobs 2');
        
        

    });

    test('my jobs button calls api to get all jobs assigned to worker and creates job cards from returned data', async () => {

        expect.assertions(7);

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        // should show job 1 and job 2
        expect(getAllAvailableJobs).toHaveBeenCalledTimes(1);
        expect(container).toHaveTextContent('testing jobs 1');
        expect(container).toHaveTextContent('testing jobs 2');
        

        // click the button
        const btn = screen.queryByTestId('my-jobs-menu-item');
        expect(btn).toBeInTheDocument();

        await act( async () => {
            getAllAssignedWorkerJobs.mockResolvedValueOnce({assignedJobs: [
                {
                    id: 1, 
                    title: 'testing jobs 1',
                    body: 'job 1 body for testing the job card',
                    postedBy: 2,
                    status: 'active',
                    assignedTo: 1,
                    applicants: []
                }
            ]});
            
            fireEvent.click(btn);
        })
        
        expect(getAllAssignedWorkerJobs).toHaveBeenCalledTimes(1);
        // should now show job 2 and not job 1
        expect(container).toHaveTextContent('testing jobs 1');
        expect(container).not.toHaveTextContent('testing jobs 2');
    });

    test('available jobs button calls api to get all available jobs for worker and creates job cards from returned data', 
        async () => {

            expect.assertions(8);

            let renderResult;

            await act(async () => {renderResult = render(
                <UserContext.Provider value={userValue}>
                    <Jobs />
                </UserContext.Provider>
            )
            });
            const { container } = renderResult;

            // should show job 1 and job 2
            expect(container).toHaveTextContent('testing jobs 1');
            expect(container).toHaveTextContent('testing jobs 2');

            // click the my jobs button
            const myJobsBtn = screen.queryByTestId('my-jobs-menu-item');
            expect(myJobsBtn).toBeInTheDocument();

            await act( async () => {
                getAllAssignedWorkerJobs.mockResolvedValueOnce({assignedJobs: [
                    {
                        id: 1, 
                        title: 'testing jobs 1',
                        body: 'job 1 body for testing the job card',
                        posted_by: 2,
                        assigned_to: 1,
                        status: 'pending',
                        applicants: []
                    }
                ]});
                fireEvent.click(myJobsBtn);
            })

            expect(getAllAssignedWorkerJobs).toHaveBeenCalledTimes(1);
            // click the available jobs button
            const availableJobsBtn = screen.queryByTestId('available-jobs-menu-item');
            expect(availableJobsBtn).toBeInTheDocument();

            await act( async () => {
                getAllAvailableJobs.mockResolvedValueOnce({allJobs: [
                    {
                        id: 1, 
                        title: 'testing jobs 1',
                        body: 'job 1 body for testing the job card',
                        posted_by: 2,
                        status: 'pending',
                        applicants: []
                    }
                ]});
                fireEvent.click(availableJobsBtn);
            })

            // should have called api to getAllAvailableJobs twice, once at the render and once on button click
            expect(getAllAvailableJobs).toHaveBeenCalledTimes(2);

            // should now show job 1 and not job 2
            expect(container).toHaveTextContent('testing jobs 1');
            expect(container).not.toHaveTextContent('testing jobs 2');
    
       
    });

});

describe('tests the createJobCards function following api calls to retrieve job data for users', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: false}});
        getActiveUserJobs.mockResolvedValueOnce({jobs: [
            {
                id: 1, 
                title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                postedBy: 1, 
                status: 'active',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'active',
                applicants: []
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });

    test('initial render for user calls api to get jobs posted by user and creates job cards from returned data', async () => {

        expect.assertions(3);

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        expect(getActiveUserJobs).toHaveBeenCalledTimes(1);
        expect(container).toHaveTextContent('testing jobs 1');
        expect(container).toHaveTextContent('testing jobs 2');
        

    });

    test('pending review button calls api to get users completed jobs and creates job cards from returned data', async () => {

        expect.assertions(6);

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;


        expect(container).toHaveTextContent('testing jobs 1');
        expect(container).toHaveTextContent('testing jobs 2');
        expect(container).not.toHaveTextContent('testing jobs 3');

        // click the button
        const button = screen.getByTestId('pending-review-menu-item');

        await act(async () => {
            getPendingReviewUserJobs.mockResolvedValueOnce({pendingReviewUserJobs: [
            {
                id: 3, 
                title: 'testing jobs 3', 
                body: 'job 3 body', 
                postedBy: 2, 
                status: 'pending review',
                applicants: []
            }]});

            fireEvent.click(button);
        })
        
        // should now show job 3 and not job 1 or 2
        expect(container).toHaveTextContent('testing jobs 3');
        expect(container).not.toHaveTextContent('testing jobs 1');
        expect(container).not.toHaveTextContent('testing jobs 2');
    });

});


describe('tests the jobCard integration for workers', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: true, applications: [2]}});
        getAllAvailableJobs.mockResolvedValueOnce({allJobs: [
            {
                id: 1, 
                title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                postedBy: 1, 
                status: 'active',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'pending',
                applicants: []
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });


    test('should show sliced job body to 30 characters', async () => {

        expect.assertions(2);

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        
        expect(container).toHaveTextContent('job 1 body for testing the job...');
        expect(container).not.toHaveTextContent('job 1 body for testing the job card');

    });

});

describe('tests the jobCard integration for users', () => {

    beforeEach(() => {
        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: false}});
        getActiveUserJobs.mockResolvedValueOnce({jobs: [
            {
                id: 1, 
                title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                posted_by: 1,
                before_image_url: 'test before image 1',
                after_image_url: '', 
                after: 'test after image', 
                status: 'active',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                posted_by: 1,
                before_image_url: 'test before image 2',
                after_image_url: '',
                status: 'active',
                applicants: []
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });


    test('should show sliced job body to 30 characters', async () => {

        expect.assertions(3);

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        // should show job 1 and not job 2
        await waitFor(() => {
            expect(container).toHaveTextContent('testing jobs 1');
        })

        expect(container).toHaveTextContent('job 1 body for testing the job...');
        expect(container).not.toHaveTextContent('job 1 body for testing the job card');

    });

    test('should render job details', async () => {

        expect.assertions(3);
        

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        // should show job 1 and not job 2
        expect(container).toHaveTextContent('testing jobs 1');
        

        const jobCard = document.querySelector('.jobCard-card')
        expect(jobCard).toBeInTheDocument();

        await act(async () => {
            getBeforeImage.mockResolvedValueOnce({preSignedUrl: 'presigned_before'});
            getActiveUserJobs.mockResolvedValueOnce({jobs: [
                {
                    id: 1, 
                    title: 'testing jobs 1', 
                    body: 'job 1 body for testing the job card', 
                    posted_by: 1,
                    before_image_url: 'test before image 1',
                    after_image_url: '', 
                    after: 'test after image', 
                    status: 'active',
                    applicants: []
                }]});
                getSingleUser.mockResolvedValueOnce({user: {id: 1, firstName: "Joe", lastName: "D.", isWorker: false}});
            // click the jobCard
            fireEvent.click(jobCard);
        })
        

        // should show job details card
        const jobDetails = screen.getByTestId('jobDetails-card');
        expect(jobDetails).toBeInTheDocument();

    });

});


describe('tests the createJob function integration', () => {

    beforeEach(() => {
        
        getActiveUserJobs.mockResolvedValueOnce({jobs: [
            {
                id: 1, title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                postedBy: 1, 
                status: 'pending',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'pending',
                applicants: []
            }
        ]});
        
    });
      
    afterEach(() => {
      jest.clearAllMocks(); // Clears all mocked function calls
    });

    test('create job button does not render for worker', async () => {

        expect.assertions(1);

        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: true}});
        getAllAvailableJobs.mockResolvedValueOnce({allJobs: [
            {
                id: 1, 
                title: 'testing jobs 1', 
                body: 'job 1 body for testing the job card', 
                postedBy: 1, 
                status: 'active',
                applicants: []
            }, 
            {
                id: 2, 
                title: 'testing jobs 2', 
                body: 'job 2 body', 
                postedBy: 1, 
                status: 'pending',
                applicants: []
            }
        ]});
        
        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        // button should not be rendered
        expect(container).not.toHaveTextContent('Create New Job');
    

    });

    test('create job button renders for user', async () => {

        expect.assertions(1);

        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: false}});

        let renderResult;

        await act(async () => {renderResult = render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });
        const { container } = renderResult;

        // button should be rendered
        expect(container).toHaveTextContent('Create New Job');
   

    });

    test('create job button reveals createJob form', async () => {

        // expect.assertions(9);

        getSingleUser.mockResolvedValueOnce({user: {id: 1, isWorker: false, applications: []}});
        getActiveUserJobs.mockResolvedValueOnce({jobs: [{}]});



        await act(async () => {render(
            <UserContext.Provider value={userValue2}>
                <Jobs />
            </UserContext.Provider>
        )
        });

        
        // the form should not be visible
        expect(screen.queryByTestId("createJob-form-container")).not.toBeInTheDocument();

        const button = screen.getByTestId("jobs-create-job-button");

        // click the create job button
        await act(async () => {
            getActiveUserJobs.mockResolvedValueOnce({jobs: [{}]});
            fireEvent.click(button);
        })

        // the form should now be visible
        expect(screen.getByTestId("createJob-form-title-input")).toBeInTheDocument();

    });

});
