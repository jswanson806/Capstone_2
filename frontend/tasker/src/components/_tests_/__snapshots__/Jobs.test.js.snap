// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`smoke and snapshot tests should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="jobs-container"
  >
    <div
      class="jobs-header-container"
    >
      <h2
        class="jobs-dynamic-page-header"
      >
        Available Jobs
      </h2>
    </div>
    <div
      class="jobs-buttons-container"
    >
      <div>
        <div
          class="btn-group"
          role="group"
        >
          <div
            class="dropdown"
          >
            <button
              aria-expanded="false"
              aria-haspopup="true"
              class="dropdown-toggle btn btn-info"
              type="button"
            >
              Sort
            </button>
            <div
              aria-hidden="true"
              class="dropdown-menu"
              role="menu"
              tabindex="-1"
            >
              <button
                class="dropdown-item"
                data-testid="available-jobs-menu-item"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                Available Jobs
              </button>
              <button
                class="dropdown-item"
                data-testid="applications-menu-item"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                Applications
              </button>
              <button
                class="dropdown-item"
                data-testid="my-jobs-menu-item"
                role="menuitem"
                tabindex="0"
                type="button"
              >
                My Jobs
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="jobs-jobCard-container"
    >
      <div
        class="row row-cols-sm-1 row-cols-md-4"
      >
        <div
          class="col"
        >
          <div
            class="col-sm-5"
          >
            <div
              class="text-center card card-body"
              style="width: 15rem;"
            >
              <div />
              <div
                class="jobCard-card"
              >
                <h5
                  class="card-title"
                >
                  testing jobs 2
                </h5>
                <div
                  class="jobCard-badge"
                >
                  <span
                    class="badge badge-success"
                    style="padding: 5px;"
                  >
                    Applied
                  </span>
                </div>
                <p
                  class="card-text"
                >
                  job 2 body...
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
